datatype/datatype.go

```diff
@@ -25,91 +25,121 @@ type DataType interface {
-	primitive.DataTypeCodeTimestamp: &primitiveTypeCodec{Timestamp},
-	primitive.DataTypeCodeUuid:      &primitiveTypeCodec{Uuid},
-	primitive.DataTypeCodeVarchar:   &primitiveTypeCodec{Varchar},
-	primitive.DataTypeCodeVarint:    &primitiveTypeCodec{Varint},
-	primitive.DataTypeCodeTimeuuid:  &primitiveTypeCodec{Timeuuid},
-	primitive.DataTypeCodeInet:      &primitiveTypeCodec{Inet},
-	primitive.DataTypeCodeDate:      &primitiveTypeCodec{Date},
-	primitive.DataTypeCodeTime:      &primitiveTypeCodec{Time},
-	primitive.DataTypeCodeSmallint:  &primitiveTypeCodec{Smallint},
-	primitive.DataTypeCodeTinyint:   &primitiveTypeCodec{Tinyint},
-	primitive.DataTypeCodeDuration:  &primitiveTypeCodec{Duration},
-	primitive.DataTypeCodeList:      &listTypeCodec{},
-	primitive.DataTypeCodeSet:       &setTypeCodec{},
-	primitive.DataTypeCodeMap:       &mapTypeCodec{},
-	primitive.DataTypeCodeTuple:     &tupleTypeCodec{},
-	primitive.DataTypeCodeUdt:       &userDefinedTypeCodec{},
-	primitive.DataTypeCodeCustom:    &customTypeCodec{},
-}
-
-func findCodec(code primitive.DataTypeCode) (codec, error) {
-	codec, ok := codecs[code]
-	if !ok {
-		return nil, fmt.Errorf("cannot find codec for data type code %v", code)
+		switch primitive.DataTypeCode(typeCode) {
+		case primitive.DataTypeCodeAscii:
+			return Ascii, nil
+		case primitive.DataTypeCodeBigint:
+			return Bigint, nil
+		case primitive.DataTypeCodeBlob:
+			return Blob, nil
+		case primitive.DataTypeCodeBoolean:
+			return Boolean, nil
+		case primitive.DataTypeCodeCounter:
+			return Counter, nil
+		case primitive.DataTypeCodeDecimal:
+			return Decimal, nil
+		case primitive.DataTypeCodeDouble:
+			return Double, nil
+		case primitive.DataTypeCodeFloat:
+			return Float, nil
+		case primitive.DataTypeCodeInt:
+			return Int, nil
+		case primitive.DataTypeCodeTimestamp:
+			return Timestamp, nil
+		case primitive.DataTypeCodeUuid:
+			return Uuid, nil
+		case primitive.DataTypeCodeVarchar:
+			return Varchar, nil
+		case primitive.DataTypeCodeVarint:
+			return Varint, nil
+		case primitive.DataTypeCodeTimeuuid:
+			return Timeuuid, nil
+		case primitive.DataTypeCodeInet:
+			return Inet, nil
+		case primitive.DataTypeCodeDate:
+			return Date, nil
+		case primitive.DataTypeCodeTime:
+			return Time, nil
+		case primitive.DataTypeCodeSmallint:
+			return Smallint, nil
+		case primitive.DataTypeCodeTinyint:
+			return Tinyint, nil
+		case primitive.DataTypeCodeDuration:
+			if version < primitive.ProtocolVersion5 {
+				return nil, fmt.Errorf("cannot use duration type with protocol version %v", version)
+			}
+			return Duration, nil
+		case primitive.DataTypeCodeCustom:
+			return readCustomType(source, version)
+		case primitive.DataTypeCodeList:
+			return readListType(source, version)
+		case primitive.DataTypeCodeMap:
+			return readMapType(source, version)
+		case primitive.DataTypeCodeSet:
+			return readSetType(source, version)
+		case primitive.DataTypeCodeUdt:
```

Comment:
Same for UDT and Tuple, they were added in v3.

Reply:
done